upf/upf_gtpu_encap.c

```diff
@@ -57,6 +57,274 @@ typedef enum
   UPF_ENCAP_N_NEXT,
 } upf_encap_next_t;
 
+static u16
+encode_ext_header_udp_port (vlib_buffer_t * b)
+{
+  u8 *p = vlib_buffer_get_current (b);
+  u16 port = clib_host_to_net_u16 (GTPU_UP_UDP_PORT);
+
+  /*
+   * TS 29.281 Clause 5.2.2.1, Length field valuse is 1. Therefore, length
+   *  of IE is 4 bytes.
+   */
+  b->current_length = (GTPU_EXT_HEADER_UPD_PORT_LENGTH * 4);
+  *p = GTPU_EXT_HEADER_UPD_PORT_LENGTH;
+  p++;
+
+  memcpy(p, &port, sizeof (u16));
+  p += sizeof (u16);
+
+  *p = GTPU_EXT_HEADER_NEXT_HEADER_NO_MORE;
+
+  return b->current_length;
+}
+
+static u16
+encode_error_indication (vlib_buffer_t * b, gtp_error_ind_t * error, int is_ip4)
+{
+  u8 *p = vlib_buffer_get_current (b);
+  u16 ip_length, nlength;
+  gtpu_ie_t teidi, recovery, gsn_address;
+
+  ip_length = (is_ip4) ? 4 : 16;
+  nlength = clib_host_to_net_u16 (ip_length);
+
+  teidi.id = GTPU_IE_TEID_I;
+  teidi.payload_len = sizeof (u32);
+
+  recovery.id = GTPU_IE_RECOVERY;
+  recovery.payload_len = sizeof (u8);
+
+  gsn_address.id = GTPU_IE_GSN_ADDRESS;
+  gsn_address.payload_len = sizeof (u16) + ip_length;
+
+  b->current_length = teidi.payload_len + gsn_address.payload_len + recovery.payload_len + (sizeof (u8) * 3);
+
+  *p = teidi.id; p++;
+  memcpy (p, &error->teid, teidi.payload_len); p += teidi.payload_len;
+
+  *p = recovery.id; p++; *p = 0; p += recovery.payload_len;
+
+  *p = gsn_address.id; p++;
+  memcpy (p, &nlength, sizeof (u16));
+  p += gsn_address.payload_len - ip_length;
+  memcpy (p, (is_ip4) ? (void *) &error->addr.ip4 : (void *) &error->addr.ip6, ip_length);
+
+  return b->current_length; //bytes encoded
+}
```

Comment:
why so complicated?
```suggestion
typedef struct
{
  u8 id;
  u8 data[];
} gtpu_tv_ie_t;

typedef struct
{
  u8 id;
  u16 len;
  u8 data[];
} gtpu_tlv_ie_t;


static u16
encode_error_indication (vlib_buffer_t * b, gtp_error_ind_t * error, int is_ip4)
{
  u8 *p = vlib_buffer_get_current (b);
  u16 ip_length, nlength;
  gtpu_tv_ie_t *tv_ie;
  gtpu_tlv_ie_t *tlv_ie;

  tv_ie = (gtpu_tv_ie_t *)p;
  tv_ie->id = GTPU_IE_TEID_I;
  *((u32 *)&tv_ie->data) = error->teid;

  p += sizeof(*tv_ie) + sizeof (u32);

  tv_ie = (gtpu_tv_ie_t *)p;
  tv_ie->id = GTPU_IE_RECOVERY;
  *((u8 *)&tv_ie->data) = 0;

  p += sizeof(*tv_ie) + sizeof (u8);

  tlv_ie = (gtpu_tlv_ie_t *)p;
  tlv_ie->id = GTPU_IE_GSN_ADDRESS;
  if (is_ip4)
    {
      tlv_ie->len = clib_host_to_net_u16(sizeof (ip4_address_t));
      clib_memcpy_fast (&tlv_ie->data, error->addr.ip4, sizeof (ip4_address_t));
    }
  else
    {
      tlv_ie->len = clib_host_to_net_u16(sizeof (ip6_address_t));
      clib_memcpy_fast (&tlv_ie->data, error->addr.ip6, sizeof (ip6_address_t));
    }

  p += sizeof(*tlv_ie) + tlv_ie->len;

  b->current_length = p - vlib_buffer_get_current (b);
  return b->current_length; //bytes encoded```

Reply:
Thanks for hints.
I based my implementation of encoders based on original `decode_error_indication_ext_hdr` and `decode_error_indication` funcs and tried to kept this logic to be compatible with existing implementation.
I've adjusted your proposals a bit.